fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubColumnsRule.java

```diff
@@ -0,0 +1,319 @@
+            LogicalScanOperator scan = optExpression.getOp().cast();
+            Set<ColumnRefOperator> scanColumns = scan.getColRefToColumnMetaMap().keySet();
+
+            ComplexColumnAccessPathNormalizer normalizer = new ComplexColumnAccessPathNormalizer();
+
+            allComplexColumns.forEach((k, v) -> {
+                if (v.containsAny(scanColumns)) {
+                    normalizer.add(k);
+                }
+            });
+
+            List<ColumnAccessPath> accessPaths = Lists.newArrayList();
+
+            for (ColumnRefOperator ref : scanColumns) {
+                if (!normalizer.hasPath(ref)) {
+                    continue;
+                }
+                String columnName = scan.getColRefToColumnMetaMap().get(ref).getName();
+                ColumnAccessPath p = normalizer.normalizePath(ref, columnName);
+
+                if (p.hashChildPath()) {
+                    accessPaths.add(p);
+                }
+            }
+
+            if (accessPaths.isEmpty()) {
+                return optExpression;
+            }
+
+            LogicalScanOperator.Builder builder = OperatorBuilderFactory.build(scan);
+            Operator newScan = builder.withOperator(scan).setColumnAccessPaths(accessPaths).build();
+            return OptExpression.create(newScan, optExpression.getInputs());
+        }
+    }
+
+    /*
+     * normalize expression to ColumnAccessPath
+     */
+    private static class ComplexColumnAccessPathNormalizer extends ScalarOperatorVisitor<Void, Void> {
+        private static final List<String> NORMALIZE_FUNCTIONS =
+                ImmutableList.of(FunctionSet.MAP_KEYS, FunctionSet.MAP_SIZE, FunctionSet.MAP_VALUES);
+
+        private final Deque<AccessPath> allAccessPaths = Lists.newLinkedList();
+
+        private AccessPath currentPath = null;
+
+        private ColumnAccessPath normalizePath(ColumnRefOperator root, String columnName) {
+            List<AccessPath> paths = allAccessPaths.stream().filter(path -> path.root().equals(root))
+                    .sorted((o1, o2) -> Integer.compare(o2.paths.size(), o1.paths.size()))
+                    .collect(Collectors.toList());
+
+            ColumnAccessPath rootPath = new ColumnAccessPath(TAccessPathType.ROOT, columnName);
+            for (AccessPath accessPath : paths) {
+                ColumnAccessPath parentPath = rootPath;
+                for (int i = 0; i < accessPath.paths.size(); i++) {
+                    if (parentPath.hasChildPath(accessPath.paths.get(i))) {
+                        parentPath = parentPath.getChildPath(accessPath.paths.get(i));
+                    } else {
+                        ColumnAccessPath childPath =
+                                new ColumnAccessPath(accessPath.pathTypes.get(i), accessPath.paths.get(i));
+                        parentPath.addChildPath(childPath);
+                        parentPath = childPath;
+                    }
+                }
+                parentPath.clearChildPath();
```

Comment:
why?

Reply:
like that： select a.b.c.d, a.b.c, a.b
we need return whole a.b, so the method will sort by path length first, then last one's children(clear a.b's children)